File: 2.6.4_slim/Downloader.java
import java.io.*;
import java.net.*;
import java.nio.file.*;
import java.security.*;
import java.util.*;
import java.util.regex.*;

public class Downloader {
    private static final String USER_AGENT = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.36";
    private static final int CONNECT_TIMEOUT = 10000; // 10 seconds
    private static final int READ_TIMEOUT = 30000;    // 30 seconds
    
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter Text or SHA256 Hash: ");
        String userText = scanner.nextLine().trim();
        
        String hash;
        if (isValidSha256(userText)) {
            hash = userText.toLowerCase();
            System.out.println("Using provided Hash: " + hash);
        } else {
            hash = generateSha256(userText);
            System.out.println("Generated Hash: " + hash);
            System.out.println("Original Text: " + userText);
        }
        
        String dataServersDir = "data_servers";
        String hashDir = dataServersDir + File.separator + hash;
        String indexFile = hashDir + File.separator + "index.html";
        String dataDir = "data";
        
        try {
            // Create directories if they don't exist
            Files.createDirectories(Paths.get(dataServersDir));
            Files.createDirectories(Paths.get(dataDir));
            
            String htmlContent = null;
            if (Files.exists(Paths.get(indexFile))) {
                System.out.println("Found cached version.");
                htmlContent = new String(Files.readAllBytes(Paths.get(indexFile)));
                
                // Download linked files
                System.out.println("Downloading linked files...");
                Map<String, List<Map<String, Object>>> downloadResults = downloadLinkedFiles(htmlContent, "", hash, dataDir);
                printDownloadResults(downloadResults);
            } else {
                String serversFile = "servers.txt";
                if (!Files.exists(Paths.get(serversFile))) {
                    System.err.println("Error: 'servers.txt' file not found.");
                    return;
                }
                
                List<String> servers = Files.readAllLines(Paths.get(serversFile));
                List<Map<String, String>> foundServers = new ArrayList<>();
                boolean contentSaved = false;
                String successfulServer = "";
                
                for (String server : servers) {
                    server = server.trim();
                    if (server.isEmpty()) continue;
                    
                    String originalServer = server;
                    String urlToCheck = buildProperUrl(server, hash, "index.html");
                    
                    StringBuilder errorMessage = new StringBuilder();
                    String pageContent = getContent(urlToCheck, errorMessage);
                    
                    if (pageContent != null) {
                        Map<String, String> serverInfo = new HashMap<>();
                        serverInfo.put("original", originalServer);
                        serverInfo.put("processed", server);
                        serverInfo.put("full_url", urlToCheck);
                        foundServers.add(serverInfo);
                        
                        if (!contentSaved) {
                            Files.createDirectories(Paths.get(hashDir));
                            successfulServer = server;
                            
                            // Process the content for local viewing
                            String processedContent = processHtmlLinks(pageContent, server, hash);
                            Files.write(Paths.get(indexFile), processedContent.getBytes());
                            contentSaved = true;
                            
                            System.out.println("File found and saved locally.");
                            System.out.println("Downloading linked files...");
                            Map<String, List<Map<String, Object>>> downloadResults = downloadLinkedFiles(pageContent, server, hash, dataDir);
                            printDownloadResults(downloadResults);
                        }
                    }
                }
                
                if (foundServers.isEmpty()) {
                    System.out.println("File not found on any server.");
                } else {
                    System.out.println("\nFile Found on These Servers:");
                    for (Map<String, String> server : foundServers) {
                        System.out.println("Server: " + server.get("original"));
                        System.out.println("Full URL: " + server.get("full_url") + "\n");
                    }
                }
            }
            
            if (Files.exists(Paths.get(indexFile))) {
                System.out.println("\nLocal HTML file: " + Paths.get(indexFile).toAbsolutePath());
            }
            
        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
            e.printStackTrace();
        }
    }
    
    private static String getContent(String url, StringBuilder errorMessage) {
        HttpURLConnection connection = null;
        try {
            URL urlObj = new URL(url);
            connection = (HttpURLConnection) urlObj.openConnection();
            connection.setRequestMethod("GET");
            connection.setRequestProperty("User-Agent", USER_AGENT);
            connection.setConnectTimeout(CONNECT_TIMEOUT);
            connection.setReadTimeout(READ_TIMEOUT);
            
            // Skip SSL verification (not recommended for production)
            if (url.startsWith("https")) {
                trustAllCertificates();
            }
            
            int responseCode = connection.getResponseCode();
            if (responseCode >= 400) {
                errorMessage.append("HTTP status code: ").append(responseCode);
                return null;
            }
            
            try (BufferedReader in = new BufferedReader(new InputStreamReader(connection.getInputStream()))) {
                StringBuilder response = new StringBuilder();
                String inputLine;
                while ((inputLine = in.readLine()) != null) {
                    response.append(inputLine).append("\n");
                }
                return response.toString();
            }
        } catch (Exception e) {
            errorMessage.append(e.getMessage());
            return null;
        } finally {
            if (connection != null) {
                connection.disconnect();
            }
        }
    }
    
    private static void trustAllCertificates() {
        // This is insecure and should not be used in production
        javax.net.ssl.HttpsURLConnection.setDefaultHostnameVerifier(
            (hostname, session) -> true);
        javax.net.ssl.TrustManager[] trustAllCerts = new javax.net.ssl.TrustManager[]{
            new javax.net.ssl.X509TrustManager() {
                public java.security.cert.X509Certificate[] getAcceptedIssuers() { return null; }
                public void checkClientTrusted(java.security.cert.X509Certificate[] certs, String authType) {}
                public void checkServerTrusted(java.security.cert.X509Certificate[] certs, String authType) {}
            }
        };
        try {
            javax.net.ssl.SSLContext sc = javax.net.ssl.SSLContext.getInstance("SSL");
            sc.init(null, trustAllCerts, new java.security.SecureRandom());
            javax.net.ssl.HttpsURLConnection.setDefaultSSLSocketFactory(sc.getSocketFactory());
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
    
    private static String buildProperUrl(String server, String hash, String filename) {
        server = server.replaceAll("/+$", "");
        return server + "/data/" + hash + "/" + filename;
    }
    
    private static String sanitizeFilename(String filename) {
        // Remove path traversal and invalid characters
        filename = filename.replaceAll(".*/", "");
        filename = filename.replaceAll("[^a-zA-Z0-9._-]", "_");
        return filename;
    }
    
    private static String getFileExtension(String url) {
        String path = url.replaceAll("[?#].*$", "");
        int lastDot = path.lastIndexOf('.');
        if (lastDot > 0) {
            return path.substring(lastDot + 1);
        }
        return "";
    }
    
    private static String getFilenameWithoutExtension(String url) {
        String path = url.replaceAll("[?#].*$", "");
        path = path.replaceAll(".*/", "");
        int lastDot = path.lastIndexOf('.');
        if (lastDot > 0) {
            return path.substring(0, lastDot);
        }
        return path.isEmpty() ? "file" : path;
    }
    
    private static boolean isValidSha256(String hash) {
        return hash.matches("^[a-f0-9]{64}$");
    }
    
    private static String generateSha256(String input) {
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] hashBytes = digest.digest(input.getBytes("UTF-8"));
            StringBuilder hexString = new StringBuilder();
            for (byte b : hashBytes) {
                String hex = Integer.toHexString(0xff & b);
                if (hex.length() == 1) hexString.append('0');
                hexString.append(hex);
            }
            return hexString.toString();
        } catch (Exception e) {
            throw new RuntimeException("Error generating SHA-256 hash", e);
        }
    }
    
    private static Map<String, List<Map<String, Object>>> downloadLinkedFiles(
            String htmlContent, String baseServer, String hash, String dataDir) throws IOException {
        List<Map<String, Object>> downloadedFiles = new ArrayList<>();
        List<Map<String, Object>> failedDownloads = new ArrayList<>();
        
        // Create data directory if it doesn't exist
        Files.createDirectories(Paths.get(dataDir));
        
        // Extract all src and href attributes
        Pattern pattern = Pattern.compile("(href|src)=[\"']([^\"']+)[\"']", Pattern.CASE_INSENSITIVE);
        Matcher matcher = pattern.matcher(htmlContent);
        
        while (matcher.find()) {
            String attribute = matcher.group(1);
            final String originalUrl = matcher.group(2);
            
            // Skip data URLs, javascript, mailto, etc.
            if (originalUrl.matches("^(data:|javascript:|mailto:|#).*")) {
                continue;
            }
            
            // Skip HTML files for href attributes (likely navigation links)
            if (attribute.equalsIgnoreCase("href") && originalUrl.matches(".*\\.html?$")) {
                continue;
            }
            
            // Convert relative URLs to absolute
            String processedUrl = originalUrl;
            if (!originalUrl.matches("^https?://")) {
                baseServer = baseServer.replaceAll("/+$", "");
                if (originalUrl.startsWith("/")) {
                    processedUrl = baseServer + originalUrl;
                } else {
                    processedUrl = baseServer + "/data/" + hash + "/" + originalUrl;
                }
            }
            
            // Get filename and create subdirectory
            String filename = sanitizeFilename(processedUrl.replaceAll(".*/", ""));
            if (filename.isEmpty() || filename.equals(".")) {
                filename = "index.html";
            }
            
            String filenameWithoutExt = getFilenameWithoutExtension(processedUrl);
            String fileDir = dataDir + File.separator + sanitizeFilename(filenameWithoutExt);
            
            Files.createDirectories(Paths.get(fileDir));
            
            String filePath = fileDir + File.separator + filename;
            
            // Skip if already downloaded
            final String finalUrl = processedUrl;
            boolean alreadyProcessed = downloadedFiles.stream()
                .anyMatch(f -> f.get("url").equals(finalUrl));
            if (alreadyProcessed) {
                continue;
            }
            
            // Check if file already exists
            if (Files.exists(Paths.get(filePath))) {
                Map<String, Object> fileInfo = new HashMap<>();
                fileInfo.put("url", processedUrl);
                fileInfo.put("local_path", filePath);
                fileInfo.put("size", Files.size(Paths.get(filePath)));
                fileInfo.put("status", "already_exists");
                downloadedFiles.add(fileInfo);
                continue;
            }
            
            // Download the file
            StringBuilder errorMessage = new StringBuilder();
            byte[] fileContent = getContentBytes(processedUrl, errorMessage);
            
            if (fileContent != null) {
                Files.write(Paths.get(filePath), fileContent);
                
                Map<String, Object> fileInfo = new HashMap<>();
                fileInfo.put("url", processedUrl);
                fileInfo.put("local_path", filePath);
                fileInfo.put("size", fileContent.length);
                fileInfo.put("status", "downloaded");
                downloadedFiles.add(fileInfo);
            } else {
                Map<String, Object> failedInfo = new HashMap<>();
                failedInfo.put("url", processedUrl);
                failedInfo.put("error", errorMessage.toString());
                failedDownloads.add(failedInfo);
            }
        }
        
        Map<String, List<Map<String, Object>>> results = new HashMap<>();
        results.put("downloaded", downloadedFiles);
        results.put("failed", failedDownloads);
        return results;
    }
    
    private static byte[] getContentBytes(String url, StringBuilder errorMessage) {
        HttpURLConnection connection = null;
        try {
            URL urlObj = new URL(url);
            connection = (HttpURLConnection) urlObj.openConnection();
            connection.setRequestMethod("GET");
            connection.setRequestProperty("User-Agent", USER_AGENT);
            connection.setConnectTimeout(CONNECT_TIMEOUT);
            connection.setReadTimeout(READ_TIMEOUT);
            
            // Skip SSL verification (not recommended for production)
            if (url.startsWith("https")) {
                trustAllCertificates();
            }
            
            int responseCode = connection.getResponseCode();
            if (responseCode >= 400) {
                errorMessage.append("HTTP status code: ").append(responseCode);
                return null;
            }
            
            try (InputStream in = connection.getInputStream();
                 ByteArrayOutputStream out = new ByteArrayOutputStream()) {
                byte[] buffer = new byte[4096];
                int bytesRead;
                while ((bytesRead = in.read(buffer)) != -1) {
                    out.write(buffer, 0, bytesRead);
                }
                return out.toByteArray();
            }
        } catch (Exception e) {
            errorMessage.append(e.getMessage());
            return null;
        } finally {
            if (connection != null) {
                connection.disconnect();
            }
        }
    }
    
    private static String processHtmlLinks(String content, String server, String hash) {
        Pattern pattern = Pattern.compile("(href|src)=[\"'](?!https?://|//|data:)([^\"']+)[\"']", Pattern.CASE_INSENSITIVE);
        Matcher matcher = pattern.matcher(content);
        StringBuffer sb = new StringBuffer();
        
        while (matcher.find()) {
            String url = matcher.group(2);
            server = server.replaceAll("/+$", "");
            
            String absoluteUrl;
            if (url.startsWith("/")) {
                absoluteUrl = server + url;
            } else {
                absoluteUrl = server + "/data/" + hash + "/" + url;
            }
            
            matcher.appendReplacement(sb, matcher.group(1) + "=\"" + absoluteUrl + "\"");
        }
        matcher.appendTail(sb);
        return sb.toString();
    }
    
    private static void printDownloadResults(Map<String, List<Map<String, Object>>> results) {
        List<Map<String, Object>> downloaded = results.get("downloaded");
        List<Map<String, Object>> failed = results.get("failed");
        
        long downloadedCount = downloaded.stream()
            .filter(f -> f.get("status").equals("downloaded"))
            .count();
        long existingCount = downloaded.stream()
            .filter(f -> f.get("status").equals("already_exists"))
            .count();
        long failedCount = failed.size();
        
        long totalSize = downloaded.stream()
            .mapToLong(f -> (Long) f.get("size"))
            .sum();
        
        System.out.println("\nDownload Statistics:");
        System.out.println("Downloaded: " + downloadedCount);
        System.out.println("Already Existed: " + existingCount);
        System.out.println("Failed: " + failedCount);
        System.out.println("Total Size: " + (totalSize / 1024) + " KB");
        
        if (!downloaded.isEmpty()) {
            System.out.println("\nDownloaded Files:");
            for (Map<String, Object> file : downloaded) {
                String status = (String) file.get("status");
                String statusMsg = status.equals("downloaded") ? "[Downloaded]" : "[Already Existed]";
                System.out.println(statusMsg + " " + file.get("url"));
                System.out.println("  -> " + file.get("local_path") + " (" + (Long) file.get("size") + " bytes)");
            }
        }
        
        if (!failed.isEmpty()) {
            System.out.println("\nFailed Downloads:");
            for (Map<String, Object> file : failed) {
                System.out.println(file.get("url") + " - Error: " + file.get("error"));
            }
        }
    }
}

File: 2.6.4_slim/downloader.php
<?php
// Function to fetch content from a URL with improved error handling
function get_content($url, &$error_message) {
    $ch = curl_init();
    curl_setopt($ch, CURLOPT_URL, $url);
    curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
    curl_setopt($ch, CURLOPT_FOLLOWLOCATION, true);
    curl_setopt($ch, CURLOPT_USERAGENT, 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.36');
    curl_setopt($ch, CURLOPT_CONNECTTIMEOUT, 10);
    curl_setopt($ch, CURLOPT_TIMEOUT, 30);
    curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, false);
    curl_setopt($ch, CURLOPT_SSL_VERIFYHOST, false);

    $data = curl_exec($ch);

    if(curl_errno($ch)) {
        $error_message = curl_error($ch);
        curl_close($ch);
        return false;
    }

    $http_code = curl_getinfo($ch, CURLINFO_HTTP_CODE);
    if ($http_code >= 400) {
        $error_message = "HTTP status code: " . $http_code;
        curl_close($ch);
        return false;
    }

    curl_close($ch);
    return $data;
}

function replace_dotdot($url) {
    return str_replace('/../', '/data/', $url);
}

function build_proper_url($server, $hash, $filename = 'index.html') {
    $server = rtrim($server, '/');
    return $server . '/data/' . $hash . '/' . $filename;
}

function sanitize_filename($filename) {
    // Remove path traversal and invalid characters
    $filename = basename($filename);
    $filename = preg_replace('/[^a-zA-Z0-9._-]/', '_', $filename);
    return $filename;
}

function get_file_extension($url) {
    $path = parse_url($url, PHP_URL_PATH);
    return pathinfo($path, PATHINFO_EXTENSION);
}

function get_filename_without_extension($url) {
    $path = parse_url($url, PHP_URL_PATH);
    $filename = pathinfo($path, PATHINFO_FILENAME);
    return $filename ? $filename : 'file';
}

function is_valid_sha256($hash) {
    return preg_match('/^[a-f0-9]{64}$/i', $hash);
}

function download_linked_files($html_content, $base_server, $hash, $data_dir) {
    $downloaded_files = [];
    $failed_downloads = [];
    
    // Create data directory if it doesn't exist
    if (!is_dir($data_dir)) {
        mkdir($data_dir, 0755, true);
    }
    
    // Extract all src and href attributes
    preg_match_all('/(href|src)=["\']([^"\']+)["\']/i', $html_content, $matches, PREG_SET_ORDER);
    
    foreach ($matches as $match) {
        $attribute = $match[1];
        $url = $match[2];
        
        // Skip data URLs, javascript, mailto, etc.
        if (preg_match('/^(data:|javascript:|mailto:|#)/', $url)) {
            continue;
        }
        
        // Skip HTML files for href attributes (likely navigation links)
        if ($attribute === 'href' && preg_match('/\.html?$/i', $url)) {
            continue;
        }
        
        // Convert relative URLs to absolute
        if (!preg_match('/^https?:\/\//', $url)) {
            $base_url = rtrim($base_server, '/');
            if (substr($url, 0, 1) === '/') {
                $url = $base_url . $url;
            } else {
                $url = $base_url . '/data/' . $hash . '/' . $url;
            }
        }
        
        // Get filename and create subdirectory
        $filename = sanitize_filename(basename(parse_url($url, PHP_URL_PATH)));
        if (empty($filename) || $filename === '.') {
            $filename = 'index.html';
        }
        
        $filename_without_ext = get_filename_without_extension($url);
        $file_dir = $data_dir . '/' . sanitize_filename($filename_without_ext);
        
        if (!is_dir($file_dir)) {
            mkdir($file_dir, 0755, true);
        }
        
        $file_path = $file_dir . '/' . $filename;
        
        // Skip if already downloaded
        if (in_array($url, array_column($downloaded_files, 'url'))) {
            continue;
        }
        
        // Check if file already exists
        if (file_exists($file_path)) {
            $downloaded_files[] = [
                'url' => $url,
                'local_path' => $file_path,
                'size' => filesize($file_path),
                'status' => 'already_exists'
            ];
            continue;
        }
        
        // Download the file
        $error_message = '';
        $file_content = get_content($url, $error_message);
        
        if ($file_content !== false) {
            if (file_put_contents($file_path, $file_content)) {
                $downloaded_files[] = [
                    'url' => $url,
                    'local_path' => $file_path,
                    'size' => strlen($file_content),
                    'status' => 'downloaded'
                ];
            } else {
                $failed_downloads[] = [
                    'url' => $url,
                    'error' => 'Failed to save file'
                ];
            }
        } else {
            $failed_downloads[] = [
                'url' => $url,
                'error' => $error_message
            ];
        }
    }
    
    return ['downloaded' => $downloaded_files, 'failed' => $failed_downloads];
}

$output_log = '';
$found_servers = [];
$download_results = [];

if (isset($_POST['user_text']) && !empty($_POST['user_text'])) {
    ob_start();

    $user_text = $_POST['user_text'];
    
    // Check if user input is already a valid SHA256 hash
    if (is_valid_sha256($user_text)) {
        $hash = strtolower($user_text);
        echo "<div class='mb-4'><strong>Using provided Hash:</strong> <code>$hash</code></div>";
    } else {
        $hash = hash('sha256', $user_text);
        echo "<div class='mb-4'><strong>Generated Hash:</strong> <code>$hash</code></div>";
        echo "<div class='info mb-4'><strong>Original Text:</strong> " . htmlspecialchars($user_text) . "</div>";
    }

    $data_servers_dir = 'data_servers';
    $hash_dir = $data_servers_dir . '/' . $hash;
    $index_file = $hash_dir . '/index.html';
    $data_dir = 'data';

    if (!is_dir($data_servers_dir)) {
        mkdir($data_servers_dir, 0755, true);
    }

    if (file_exists($index_file)) {
        echo "<div class='success mb-4'>Found cached version.</div>";
        
        // Still download linked files if they don't exist
        $html_content = file_get_contents($index_file);
        $download_results = download_linked_files($html_content, '', $hash, $data_dir);
        
    } else {
        $servers_file = 'servers.txt';
        if (!file_exists($servers_file)) {
            die("Error: 'servers.txt' file not found.");
        }

        $servers = file($servers_file, FILE_IGNORE_NEW_LINES | FILE_SKIP_EMPTY_LINES);
        $content_saved = false;
        $original_content = '';
        $successful_server = '';

        foreach ($servers as $server) {
            $server = trim($server);
            $original_server = $server;
            $url_to_check = build_proper_url($server, $hash, 'index.html');

            $curl_error = null;
            $page_content = get_content($url_to_check, $curl_error);

            if ($page_content !== false) {
                $found_servers[] = [
                    'original' => $original_server,
                    'processed' => $server,
                    'full_url' => $url_to_check
                ];

                if (!$content_saved) {
                    if (!is_dir($hash_dir)) {
                        mkdir($hash_dir, 0755, true);
                    }

                    $original_content = $page_content;
                    $successful_server = $server;

                    // Process the content for local viewing
                    $page_content = preg_replace_callback(
                        '/(href|src)=["\'](?!https?:\/\/|\/\/|data:)([^"\']+)["\']/i',
                        function($matches) use ($server, $hash) {
                            $url = $matches[2];
                            $base_url = rtrim($server, '/');
                            if (substr($url, 0, 1) === '/') {
                                $absolute_url = $base_url . $url;
                            } else {
                                $absolute_url = $base_url . '/data/' . $hash . '/' . $url;
                            }
                            return $matches[1] . '="' . $absolute_url . '"';
                        },
                        $page_content
                    );

                    file_put_contents($index_file, $page_content);
                    $content_saved = true;
                    
                    echo "<div class='success mb-4'>File found and saved locally.</div>";
                    
                    // Download all linked files
                    echo "<div class='info mb-4'>Downloading linked files...</div>";
                    $download_results = download_linked_files($original_content, $server, $hash, $data_dir);
                }
            }
        }

        if (empty($found_servers)) {
            echo "<div class='error'>File not found on any server.</div>";
        }
    }

    $output_log = ob_get_clean();
}
?>

<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Downloader</title>
    <style>
        * {
            margin: 0;
            padding: 0;
            box-sizing: border-box;
        }
        
        body {
            font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, sans-serif;
            line-height: 1.6;
            color: #333;
            min-height: 100vh;
            display: flex;
            flex-direction: column;
            align-items: center;
            justify-content: flex-start;
            padding: 20px;
            background: #f8fafc;
        }
        
        .container {
            width: 100%;
            max-width: 1000px;
            display: flex;
            flex-direction: column;
            align-items: center;
        }
        
        h1, h2, h3 {
            font-weight: 600;
            color: #111;
            text-align: center;
        }
        h1 { 
            font-size: 1.8rem; 
            margin-bottom: 2rem; 
        }
        h2 { 
            font-size: 1.4rem; 
            margin: 1.5rem 0 1rem 0; 
        }
        h3 { 
            font-size: 1.2rem; 
            margin: 1rem 0 0.5rem 0; 
        }
        
        .form-container {
            background: white;
            padding: 2rem;
            border-radius: 8px;
            box-shadow: 0 2px 10px rgba(0, 0, 0, 0.1);
            width: 100%;
            max-width: 600px;
            margin-bottom: 2rem;
        }
        
        .form-group {
            margin-bottom: 1.5rem;
            text-align: center;
        }
        
        label {
            display: block;
            margin-bottom: 1rem;
            font-weight: 500;
            color: #374151;
        }
        
        .input-container {
            display: flex;
            gap: 0.5rem;
            align-items: center;
            justify-content: center;
        }
        
        input[type="text"] {
            flex: 1;
            max-width: 400px;
            padding: 0.75rem;
            border: 1px solid #ddd;
            border-radius: 4px;
            font-family: inherit;
            font-size: 1rem;
        }
        
        input[type="text"]:focus {
            outline: none;
            border-color: #2563eb;
            box-shadow: 0 0 0 3px rgba(37, 99, 235, 0.1);
        }
        
        .btn {
            display: inline-block;
            background: #2563eb;
            color: white;
            padding: 0.75rem 1.5rem;
            border: none;
            border-radius: 4px;
            text-decoration: none;
            font-weight: 500;
            cursor: pointer;
            font-size: 1rem;
            white-space: nowrap;
        }
        .btn:hover {
            background: #1d4ed8;
        }
        .btn-success {
            background: #059669;
            margin-top: 1rem;
        }
        .btn-success:hover {
            background: #047857;
        }
        
        .tip-text {
            color: #6b7280;
            font-size: 0.9em;
            margin-top: 1rem;
            text-align: center;
        }
        
        .results {
            width: 100%;
            max-width: 1000px;
        }
        
        .success {
            color: #065f46;
            background: #d1fae5;
            padding: 1rem;
            border-radius: 4px;
            margin-bottom: 1rem;
            text-align: center;
        }
        .error {
            color: #b91c1c;
            background: #fee2e2;
            padding: 1rem;
            border-radius: 4px;
            margin-bottom: 1rem;
            text-align: center;
        }
        .info {
            color: #1e40af;
            background: #dbeafe;
            padding: 1rem;
            border-radius: 4px;
            margin-bottom: 1rem;
            text-align: center;
        }
        .servers-list {
            margin-top: 2rem;
            background: white;
            padding: 1.5rem;
            border-radius: 8px;
            box-shadow: 0 2px 10px rgba(0, 0, 0, 0.1);
        }
        .server-item {
            padding: 0.75rem;
            border: 1px solid #e5e7eb;
            border-radius: 4px;
            margin-bottom: 0.5rem;
            background: #f9fafb;
        }
        .server-url {
            font-family: Menlo, Monaco, Consolas, monospace;
            font-size: 0.9em;
            color: #1f2937;
            word-break: break-all;
        }
        .downloads-section {
            margin-top: 2rem;
            padding: 1.5rem;
            border: 1px solid #e5e7eb;
            border-radius: 8px;
            background: white;
            box-shadow: 0 2px 10px rgba(0, 0, 0, 0.1);
        }
        .download-item {
            padding: 0.5rem;
            border-bottom: 1px solid #e5e7eb;
            display: flex;
            justify-content: space-between;
            align-items: center;
        }
        .download-item:last-child {
            border-bottom: none;
        }
        .download-url {
            font-family: Menlo, Monaco, Consolas, monospace;
            font-size: 0.85em;
            color: #374151;
            flex: 1;
            margin-right: 1rem;
            word-break: break-all;
        }
        .download-size {
            font-size: 0.8em;
            color: #6b7280;
            white-space: nowrap;
        }
        .failed-item {
            background: #fef2f2;
            color: #b91c1c;
        }
        .already-exists {
            background: #fef7cd;
        }
        .status-badge {
            font-size: 0.75em;
            padding: 0.2rem 0.5rem;
            border-radius: 12px;
            margin-left: 0.5rem;
            font-weight: 500;
        }
        .status-badge.downloaded {
            background: #d1fae5;
            color: #065f46;
        }
        .status-badge:not(.downloaded) {
            background: #fef3c7;
            color: #92400e;
        }
        code {
            font-family: Menlo, Monaco, Consolas, monospace;
            background: #f3f4f6;
            padding: 0.2rem 0.4rem;
            border-radius: 3px;
            font-size: 0.9em;
        }
        .stats {
            display: flex;
            gap: 2rem;
            margin: 1rem 0;
            justify-content: center;
            flex-wrap: wrap;
        }
        .stat-item {
            text-align: center;
        }
        .stat-number {
            font-size: 1.5rem;
            font-weight: 600;
            color: #2563eb;
        }
        .stat-label {
            font-size: 0.9rem;
            color: #6b7280;
        }
        
        .copyright {
            margin-top: auto;
            padding-top: 2rem;
            text-align: center;
            color: #6b7280;
            font-size: 0.9rem;
        }
        
        .open-file-container {
            text-align: center;
            margin-top: 2rem;
        }
        
        @media (max-width: 600px) {
            .input-container {
                flex-direction: column;
                gap: 1rem;
            }
            
            input[type="text"] {
                max-width: 100%;
            }
            
            .stats {
                gap: 1rem;
            }
        }
    </style>
</head>
<body>
    <div class="container">
        <h1>Downloader</h1>
        
        <div class="form-container">
            <form action="" method="post">
                <div class="form-group">
                    <label for="user_text">Enter Text or SHA256 Hash:</label>
                    <div class="input-container">
                        <input type="text" id="user_text" name="user_text" required placeholder="Enter any text to generate SHA256 hash, or paste an existing SHA256 hash" value="<?= htmlspecialchars($_POST['user_text'] ?? '') ?>">
                        <button type="submit" class="btn">Check Servers</button>
                    </div>
                    <div class="tip-text">
                        💡 Tip: You can enter either plain text (will be hashed) or a valid SHA256 hash (will be used directly)
                    </div>
                </div>
            </form>
        </div>

        <?php if ($_SERVER['REQUEST_METHOD'] === 'POST'): ?>
            <div class="results">
                <?= $output_log ?>
                
                <?php if (!empty($found_servers)): ?>
                    <div class="servers-list">
                        <h2>File Found on These Servers:</h2>
                        <?php foreach ($found_servers as $server): ?>
                            <div class="server-item">
                                <div><strong>Server:</strong> <?= htmlspecialchars($server['original']) ?></div>
                                <div class="server-url"><strong>Full URL:</strong> <?= htmlspecialchars($server['full_url']) ?></div>
                            </div>
                        <?php endforeach; ?>
                    </div>
                <?php endif; ?>
                
                <?php if (isset($index_file) && file_exists($index_file)): ?>
                    <div class="open-file-container">
                        <a href="<?= htmlspecialchars($index_file) ?>" target="_blank" class="btn btn-success">Open Local HTML File</a>
                    </div>
                <?php endif; ?>
                
                <?php if (!empty($download_results)): ?>
                    <div class="downloads-section">
                        <h2>Downloaded Files</h2>
                        
                        <div class="stats">
                            <div class="stat-item">
                                <div class="stat-number"><?= count(array_filter($download_results['downloaded'], function($f) { return $f['status'] === 'downloaded'; })) ?></div>
                                <div class="stat-label">Downloaded</div>
                            </div>
                            <div class="stat-item">
                                <div class="stat-number"><?= count(array_filter($download_results['downloaded'], function($f) { return $f['status'] === 'already_exists'; })) ?></div>
                                <div class="stat-label">Already Existed</div>
                            </div>
                            <div class="stat-item">
                                <div class="stat-number"><?= count($download_results['failed']) ?></div>
                                <div class="stat-label">Failed</div>
                            </div>
                            <div class="stat-item">
                                <div class="stat-number"><?= number_format(array_sum(array_column($download_results['downloaded'], 'size')) / 1024, 1) ?> KB</div>
                                <div class="stat-label">Total Size</div>
                            </div>
                        </div>
                        
                        <?php if (!empty($download_results['downloaded'])): ?>
                            <h3>File Status:</h3>
                            <?php foreach ($download_results['downloaded'] as $file): ?>
                                <div class="download-item <?= $file['status'] === 'already_exists' ? 'already-exists' : '' ?>">
                                    <div class="download-url">
                                        <?= htmlspecialchars($file['url']) ?>
                                        <?php if ($file['status'] === 'already_exists'): ?>
                                            <span class="status-badge">Already Exists</span>
                                        <?php else: ?>
                                            <span class="status-badge downloaded">Downloaded</span>
                                        <?php endif; ?>
                                    </div>
                                    <div class="download-size"><?= number_format($file['size'] / 1024, 1) ?> KB</div>
                                </div>
                            <?php endforeach; ?>
                        <?php endif; ?>
                        
                        <?php if (!empty($download_results['failed'])): ?>
                            <h3>Failed Downloads:</h3>
                            <?php foreach ($download_results['failed'] as $file): ?>
                                <div class="download-item failed-item">
                                    <div class="download-url"><?= htmlspecialchars($file['url']) ?></div>
                                    <div class="download-size"><?= htmlspecialchars($file['error']) ?></div>
                                </div>
                            <?php endforeach; ?>
                        <?php endif; ?>
                    </div>
                <?php endif; ?>
            </div>
        <?php endif; ?>
        
        <div class="copyright">
            All rights reserved
        </div>
    </div>
</body>
</html>